### ISSUE TITLE
tf.contrib.learn estimators that use feature columns via the input_fun not compatible with x inputs when predicting and not exportable to serving

### ISSUE THREAD

**Issue by fventer** [tensorflow/tensorflow#5598]
I have run into into performance and portability issues with tf.conrtib.learn estimators that use high level feature columns and have done extensive code traversal to figure out a possible solution to make predictions faster.

At the core of the problem is that high level feature column tensors created in the input_fn are transformed and added as input tensors to the actual neural network and/or logistic model. After that the hidden layers and logit (output) layer is added to the graph and the logits for the linear model (if used) are also created. This happens every time you run predict on the estimator, at the cost an overhead of 3-7 seconds.

One solution I tried is based on posts that I read about the x parameter of the predict method (vs input_fn). If you create an iterator that feeds examples (that never sends the StopIteration exception) then you could keep the model "resident" avoiding the graph being built every time you call predict.

However, the x parameter as iterable does not work,  because it seems that you can only provide it low level inputs (such as numpy arrays or pandas dataframe of type float or int only). It seems that this solution precludes the nice feature columns such as sparse tensors (for categoricals) and embeddings (to get inputs to the neural net), etc.

Another problem that arises with the above issues with the estimator design is that although it has an export method to create a model that tensorflow serving loads nicely, it will not work, because the graph needs to be built at predict time to extract model inputs from the higher level feature column tensors and of course tensor flow serving does not work like that. It loads a graph and runs inputs to create outputs - similar to the predict method using the x parameter. It does not create graphs on the fly every time a predict call is made.

What is needed as a new feature is that the x and y parameters of the predict method should accept raw inputs (like strings for categorical features/embeddings) that is then passed into the feed_dict parameter of the session run call.

Similarly, for tensorflow serving, it should be possible to export the estimator and write the model client so that the raw feature column values (strings) can be sent to the model instantiated in tensorflow serving. 

This feature is quite urgent in my opinion as models based on the cool abstractions available in the contrib.learn estimator API are not very useful beyond training and evaluation without this requested feature.

**Comment by girving** [comment]
@martinwicke I don't quite understand this one.  Does it make sense to you?

**Comment by martinwicke** [comment]
I'm not entirely sure what feature you are asking for, but it is clear that there are shortcomings around the sklearn interface, `input_fn`, and how they mix, especially where serving is concerned. 

@davidsoergel is working on an update to export that I believe should fix most of your problems there.

We are discussing changes to the predict interface at the moment, including possibly reusing the graph for multiple queries, and providing a utility to run exported models without starting up a full tensorflow/serving instance.

I will close this bug, not because your complaints are invalid, but because I see no actionable feature request (beyond, "Please fix this, it's no good", which we're working on).

**Comment by fventer** [comment]
Thanks Martin

Looking forward to the fix.

**Comment by fventer** [comment]
How to I follow the updates that @davidsoergel is working on that you said for fix my mentioned problems?

**Comment by nubbel** [comment]
@fventer Apparently, this landed in master as of 7942860247b2f945e42092d43b29d9641ad2a217.
The exported *SavedModel* (similar to SessionBundle) can then be served using tensorflow/serving with the `--use_saved_model=true` option. 
I can confirm it's working great!

Thanks @davidsoergel! 🎉

**Comment by fventer** [comment]
Thanks!

I will be testing it asap.

Regards

Fritz Venter
fritz.venter@gmail.com
cell: +15122937896

**Comment by sendit2me** [comment]
@nubbel  I see you have the SavedModel working, did you get the export working with Sparse based tensors ? specifically from the Deep and Wide examples, namely sparse_column_with_keys or sparse_column_with_hash_bucket, I can get an export working if I use real_valued_column and bucketized_column (using real_valued_columns), has to do with the resulting 'SparseTensor' object has no attribute 'name'

Peter

**Comment by nubbel** [comment]
@sendit2me Hi Peter, I ran in the exact same problem.
I managed to work around it by converting the `VarLenFeature`s to `FixedLenFeature`s:
```python
def fix_feature_spec(feature_spec):
    for key, feature in feature_spec.items():
        if isinstance(feature, tf.VarLenFeature):
            feature_spec[key] = tf.FixedLenFeature(shape=[1], dtype=feature.dtype, default_value=None)
```

I also share my jupyter notebook with you where I build a model (based on the wide and deep tutorial) and export it to a `SavedModel`: https://gist.github.com/nubbel/a26b6cbb21c34bc36abbe76b850f5541
(sorry for the verbose debug logs and deprecation warnings, don't have the time for cleaning this up right now)

Hope that helps!

**Comment by sendit2me** [comment]
Thanks, that works a charm, for exporting... and loading to Tensorflow Serving...
but am now trying to figure out how to get the client format correct, did you compile a tensorflow client, I was trying to use "tensorflow-serving-python" which has the protos but this does not seem to work... I have tried a number of formats. any ideas what is wrong, or should I go straight to a tf client...

**Comment by nubbel** [comment]
I've build a [Ruby client library](https://github.com/nubbel/tensorflow_serving_client-ruby), because I need to perform predictions from a Rails app.

My client code looks like that:
```ruby
req = Tensorflow::Serving::PredictRequest.new
req.model_spec = Tensorflow::Serving::ModelSpec.new name: 'targeting_deep'

example = Tensorflow::Example.new features: Tensorflow::Features.new
example.features.feature['age'] = Tensorflow::Feature.new float_list: Tensorflow::FloatList.new(value: [customer.age || -1])
example.features.feature['gender'] = Tensorflow::Feature.new bytes_list: Tensorflow::BytesList.new(value: [customer.gender || 'n/a'])
example.features.feature['state'] = Tensorflow::Feature.new bytes_list: Tensorflow::BytesList.new(value: [customer.gender || 'n/a'])
example.features.feature['purchase_recency'] = Tensorflow::Feature.new float_list: Tensorflow::FloatList.new(value: [customer.purchase_recency || -1])
example.features.feature['purchase_frequency'] = Tensorflow::Feature.new float_list: Tensorflow::FloatList.new(value: [customer.purchase_frequency || -1])
example.features.feature['monetary_value'] = Tensorflow::Feature.new float_list: Tensorflow::FloatList.new(value: [customer.monetary_value || -1])
example.features.feature['interaction_recency'] = Tensorflow::Feature.new float_list: Tensorflow::FloatList.new(value: [customer.interaction_recency || -1])
example.features.feature['interaction_frequency'] = Tensorflow::Feature.new float_list: Tensorflow::FloatList.new(value: [customer.interaction_frequency || -1])

serialized_example = Tensorflow::Example.encode example

req.inputs['inputs'] = Tensorflow::TensorProto.new(
  string_val: [serialized_example],
  tensor_shape: Tensorflow::TensorShapeProto.new(dim: [Tensorflow::TensorShapeProto::Dim.new(size: 1)]),
  dtype: Tensorflow::DataType::DT_STRING
)

stub = Tensorflow::Serving::PredictionService::Stub.new('model-server:8500', :this_channel_is_insecure)
res = stub.predict req
res.outputs
```

It should be straightforward to translate that to python.

**Comment by jasonhere** [comment]
@nubbel Thank you for your tutorial and  you ruby client. 

Could you please offer a java version predict client since so many issues and stackoverflow questions on this.
